File: scala/spiral-matrix/src/main/scala/SpiralMatrix.scala
object SpiralMatrix {
  def spiralMatrix(size: Int): List[List[Int]] =
    List.range(0, size).map(row =>
      List.range(0, size).map(col => spiralMatrixValue(size, 1, row, col)))

  private def spiralMatrixValue(size: Int, startValue: Int, row: Int, col: Int): Int = {
    (row, col) match {
      case (0, _) => startValue + col
      case _ if col == size - 1 => startValue + size + row - 1
      case _ if row == size - 1 => startValue + size * 3 - 3 - col
      case (_, 0) => startValue + size * 4 - 4 - row
      case _ => spiralMatrixValue(size - 2, startValue + size * 4 - 4, row - 1, col -1)
    }
  }
}

File: clojure/spiral-matrix/src/spiral_matrix.clj
(ns spiral-matrix)

(defn spiral-value [size start-value row col]
  (cond
    (zero? row) (+ start-value col)
    (= col (dec size)) (+ start-value size row -1)
    (= row (dec size)) (+ start-value (* size 3) -3 (- col))
    (zero? col) (+ start-value (* size 4) -4 (- row))
    :else (recur (- size 2) (+ start-value (* size 4) -4) (dec row) (dec col))))

(defn spiral [n]
  (for [row (range 0 n)]
    (map #(spiral-value n 1 row %) (range 0 n))))
